pandas/core/arrays/array_.py

```diff
@@ -229,7 +229,7 @@ def array(data: Sequence[object],
         dtype = registry.find(dtype) or dtype
 
     if is_extension_array_dtype(dtype):
-        cls = dtype.construct_array_type()
+        cls = cast(ExtensionDtype, dtype).construct_array_type()
```

Comment:
Do know of an alternate way that `is_extension_array_dtype` would only cause type checking to expect the ExtensionDtype here rather than explicitly casting? May ultimately have applications in other pieces of codebase if possible

Reply:
I'm haven't been able to find a better solution.

Mypy infers the narrowing of Unions based on some logic, but I suspect the control flow here is more complicated than mypy can follow without help. E.g., if it were sufficient to replace `is_extension_array_dtype` here with `isinstance(dtype, ExtensionDtype)`, mypy would follow it, and we wouldn't need an explicit cast. But `is_extension_array_dtype` is more complicated than that. Even changing `is_extension_array_dtype` to simply return `isinstance(dtype, ExtensionDtype)` is too complex for mypy to follow.

Reply:
OK thanks. This looks related to https://github.com/python/mypy/issues/5206 (may not completely resolve)